### PR TITLE
Add ShardCluster implementation

### PR DESCRIPTION
**Pull request by congqixia**
ShardCluster maintains shard replica meta information
It watches node & segment change events
Provides shard replica search/query services

Related #16359 
/kind enhancement

Signed-off-by: Congqi Xia <congqi.xia@zilliz.com>

### CODE REVIEW

File: internal/querynode/shard_cluster.go [hunk 1]
```diff
@@ -0,0 +1,477 @@
...
+}
+
+func (sc *ShardCluster) getSegment(segmentID int64) (*shardSegmentInfo, bool) {
+	sc.mut.RLock()
+	defer sc.mut.RUnlock()
+	segment, ok := sc.segments[segmentID]
+	return segment, ok
+}
+
+// segmentAllocations returns segment to node mappings.
```
Comment (letian-jiang) [inline internal/querynode/shard_cluster.go:356]: I guess it should be node to segments mapping.

File: internal/querynode/shard_cluster.go [hunk 2]
```diff
@@ -0,0 +1,477 @@
...
+	}
+
+	if sc.vchannelName != req.GetDmlChannel() {
+		return nil, fmt.Errorf("ShardCluster for %s does not match to request channel :%s", sc.vchannelName, req.GetDmlChannel())
+	}
+
+	// get node allocation
+	segAllocs := sc.segmentAllocations()
+
+	// TODO dispatch to local queryShardService query dml channel growing segments
```
[the remark this answers is not in the document]
Comment (congqixia) [reply internal/querynode/shard_cluster.go:381]: @letian-jiang it depends on how we combine queryservice and shardCluster. Put this TODO as a reminder.

File: internal/querynode/shard_cluster.go (hunk 1 above)
Comment (congqixia) [reply internal/querynode/shard_cluster.go:356]: @letian-jiang nice catch. Will fix in next pr.

**Pull request merged**